### PR TITLE
feat: add VictoriaLogs log provider

### PR DESCRIPTION
**Pull request by bellorr**
Adds a new \`VictoriaLogsProvider\` that enables blocky-ui to read DNS query logs directly from a [Victoria Logs](https://docs.victoriametrics.com/victorialogs/) instance.

## Motivation

Many setups ship blocky's stdout logs to Victoria Logs via fluent-bit, Vector, or similar. When blocky runs with \`queryLog.type: console\`, it emits structured JSON to stdout — all the fields blocky-ui needs (\`client_ip\`, \`client_names\`, \`question_name\`, \`response_type\`, etc.) are already present as top-level fields in VL after JSON parsing. This provider enables the full dashboard experience (query history, top domains, top clients, queries over time) without requiring a separate database.

## Usage

\`\`\`env
QUERY_LOG_TYPE=console
QUERY_LOG_CONSOLE_PROVIDER=victorialogs
QUERY_LOG_TARGET=http://your-victoria-logs-host:9428
\`\`\`

**Requirements:**
- blocky configured with \`queryLog.type: console\` (the default)
- A log shipper (fluent-bit, Vector, etc.) configured to forward blocky's stdout to VL with JSON fields parsed at the top level

## Implementation notes

- VL 1.48.0 does not support \`count(if(...))\` multi-stat syntax — blocked/cached counts use separate parallel queries merged in JS
- Pagination for \`getQueryLogs\` is in-memory (fetch \`limit + offset\`, slice) since VL's query API has no \`OFFSET\` param
- Conformance tests use \`GenericContainer\` (testcontainers) with a pinned \`v1.48.0-victorialogs\` image — Docker required to run tests

## Preview image

Built from this branch (linux/amd64) — for testing only, not for production use:

\`\`\`
ghcr.io/bellorr/blocky-ui:victorialogs-preview
\`\`\`

> Rebuilt after rebasing with upstream main (includes MySQL pagination improvements, React compiler, dependency updates from PRs #272–#288).

## Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/GabeDuarteM/blocky-ui/blob/main/CONTRIBUTING.md)
- [x] \`bun run verify\` passes
- [x] \`bun run test\` passes (all 327 tests including the new victorialogs conformance suite)
- [x] A changeset has been added

## Why this change?

I run blocky in Kubernetes and already ship its logs to Victoria Logs via fluent-bit. All the fields blocky-ui needs are already indexed in VL as structured top-level fields — adding this provider means I get the full query history dashboard without standing up a separate database.

## AI Assistance

- [x] AI was used in this PR

Implemented with [Claude](https://claude.ai) (Anthropic). The VL query syntax was validated against a live Victoria Logs 1.48.0 instance before writing the implementation, and the full conformance test suite was used to verify correctness. I reviewed the design, validated the field mappings against my actual log data, and confirmed the complete dashboard works end-to-end in my cluster before opening this PR.

### CODE REVIEW

**Review by GabeDuarteM — COMMENTED** [review]
Hey 👋  thank you for this! more provider support is always good to have 😄

**Comment by bellorr** [comment]
@GabeDuarteM 

Thanks for the thorough feedback!

**Cross-provider consistency (`providerNames`)**

VictoriaLogs was initially excluded from the cross-provider suite because I expected timestamp/time-bucketing behavior to differ between VL's native time grouping and the SQL providers — and didn't want that to cause spurious failures. After actually running the suite with VL included, the timestamp handling turned out to be fine; the real issue was sort tie-breaking. When two rows share the same count, MySQL resolves ties case-insensitively (so `Alpha` and `alpha` are equivalent), while VictoriaLogs sorts by byte value (uppercase `A–Z` comes before lowercase `a–z`). The same seed data can produce a different order in that edge case even though both providers are correct.

The fix is to normalize both sides before comparing rather than excluding VL from the suite. The tests now define `sortDomains`/`sortClients` helpers that apply an identical secondary sort (count desc, then name asc case-insensitive via `localeCompare`) to both the baseline and the current provider result before asserting equality. VictoriaLogs is now included in `providerNames` for all cross-provider consistency checks.

---

**Env var semantics**

Updated exactly as suggested:
```env
QUERY_LOG_TYPE=console
QUERY_LOG_CONSOLE_PROVIDER=victorialogs
QUERY_LOG_TARGET=https://victoria-logs:1234
```
`QUERY_LOG_TYPE=console` mirrors blocky's own `queryLog.type: console`, and `QUERY_LOG_CONSOLE_PROVIDER` selects the backend — leaving room for Loki or others later.

---

**Documentation**

Added:
- `.env.example` — VictoriaLogs config block
- `README.md` — features list updated, env var table includes `QUERY_LOG_CONSOLE_PROVIDER`, docker-compose example includes the VictoriaLogs variant

File: src/server/logs/__tests__/provider-conformance.test.ts [hunk 1]
```diff
@@ -812,14 +814,37 @@ describe("cross-provider consistency", () => {
...
-      for (let j = 0; j < baseline.items.length; j++) {
-        expect(current.items[j]?.count).toBe(baseline.items[j]?.count);
-        expect(current.items[j]?.blocked).toBe(baseline.items[j]?.blocked);
-        expect(current.items[j]?.percentage).toBeCloseTo(
-          baseline.items[j]?.percentage ?? 0,
+      for (let j = 0; j < sortedBaseline.length; j++) {
+        expect(sortedCurrent[j]?.count).toBe(sortedBaseline[j]?.count);
+        expect(sortedCurrent[j]?.blocked).toBe(sortedBaseline[j]?.blocked);
+        expect(sortedCurrent[j]?.percentage).toBeCloseTo(
+          sortedBaseline[j]?.percentage ?? 0,
```
Comment (GabeDuarteM) [inline src/server/logs/__tests__/provider-conformance.test.ts]: is this necessary? I ran it locally without it and it appeared to pass regardless... same for below

File: src/server/logs/victorialogs/provider.ts [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    if (filter === "blocked") {
+      const rows = await this.queryRaw(
+        `${blockedBase} | stats by (question_name) count() as count | sort by (count desc)`,
+        { start },
+      );
+      const totalQueries = rows.reduce((s, r) => s + Number(r.count), 0);
+      return {
+        totalCount: rows.length,
+        items: rows.slice(offset, offset + limit).map((r) => {
```
Comment (GabeDuarteM) [inline src/server/logs/victorialogs/provider.ts]: Above we are querying it from VL and then paginating it in node, can we paginate it directly on the query instead and do a separate query to get the count? We do this for the other providers, its usually better for perf...

Also, I noticed both in here and in the other queries, we're generally ordering by only a single field, while on the other providers we also add a secondary sort as a tie-breaker. Is this a VL query limitation? Can we add the same to the VL queries?

Same on the blocked query below and other queries below

File: src/server/logs/victorialogs/provider.ts [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+      .filter((r) => r._time != null)
+      .map((r) => {
+        const time = r._time!;
+        return {
+          time,
+          total: Number(r.total),
+          blocked: blockedByTime.get(time) ?? 0,
+          cached: cachedByTime.get(time) ?? 0,
+        };
+      });
```
[the remark this answers is not in the document]
Comment (GabeDuarteM) [reply src/server/logs/victorialogs/provider.ts]: this comment from coderabbit might be actually valid, haven't tested it on the charts, but its currently different from how the other providers behave

**Review by GabeDuarteM — COMMENTED** [review]
Looks good overall, just a few comments. Most are about keeping VictoriaLogs aligned with the behavior of the other providers. I called out the specific cases I noticed below, but it would be worth comparing the rest of the queries as well to make sure they stay more or less consistent with what the existing providers do...

File: src/server/logs/__tests__/provider-conformance.test.ts [hunk 4]
```diff
@@ -726,8 +742,13 @@ defineProviderTests("mysql");
 defineProviderTests("postgres");
 defineProviderTests("csv");
 defineProviderTests("csv-client");
+defineProviderTests("victorialogs");
 
 describe("cross-provider consistency", () => {
+  // VictoriaLogs is intentionally excluded from cross-provider consistency tests.
+  // It is seeded identically but aggregation results (timestamps, sort stability
+  // for ties) may differ subtly from SQL providers. Correctness is verified by
+  // the dedicated defineProviderTests("victorialogs") suite above.
```
[the remark this answers is not in the document]
Comment (bellorr) [reply src/server/logs/__tests__/provider-conformance.test.ts]: Yes — VL is fully included, no provider is excluded. The normalisation is necessary because of a genuine tie-breaking difference between provider types: SQL and VL sort tied entries by byte value (uppercase < lowercase), while the CSV providers use JavaScript's `localeCompare` (case-insensitive). In the 30d seed data, `GitHub.com` and `example.org` both appear 7 times. Byte-order puts `GitHub.com` first (G=71 < e=101), but `localeCompare` puts `example.org` first (e < g). All providers agree on the counts and blocked values — only the tie-breaking position differs. Re-sorting both sides with the same comparator before asserting lets us confirm the data is consistent without masking any real discrepancies. Updated the comment in 74846cd to spell this out with the concrete example.

File: src/env.js [hunk 5]
```diff
@@ -9,7 +9,14 @@ export const env = createEnv({
...
     DATABASE_URL: z.string().optional().meta({ deprecated: true }),
     QUERY_LOG_TYPE: z
-      .enum(["mysql", "postgresql", "timescale", "csv", "csv-client"])
+      .enum([
+        "mysql",
+        "postgresql",
+        "timescale",
+        "csv",
+        "csv-client",
+        "victorialogs",
```
[the remark this answers is not in the document]
Comment (bellorr) [reply src/env.js]: Done — updated in commit fe17900. `QUERY_LOG_TYPE` now accepts `console` as the value (matching Blocky's own log type), and `QUERY_LOG_CONSOLE_PROVIDER=victorialogs` is a separate env var that selects the backend. This also sets things up nicely for adding Loki or other console backends later without changing the type enum.

[the remark this answers is not in the document]
Comment (bellorr) [reply src/env.js]: Added in commits 0a34346 and f2ab0c7. The README now has a full example block for the console/VictoriaLogs setup and a table entry for `QUERY_LOG_CONSOLE_PROVIDER` marked as `Conditional` (required when `QUERY_LOG_TYPE=console`). The `.env.example` has all three vars documented as well.

File: src/server/logs/__tests__/provider-conformance.test.ts (hunk 1 above)
Comment (bellorr) [reply src/server/logs/__tests__/provider-conformance.test.ts]: Yes, it's necessary. The seed data contains a real tie at count=7 between `GitHub.com` and `example.org`. VL and SQL sort by byte value (`G`=71 < `e`=101, so `GitHub.com` comes first), while the CSV providers use `localeCompare` which is case-insensitive (`e` < `g`, so `example.org` comes first). If you ran it locally without the normalisation and it passed, you likely had a different dataset or the tie didn't surface — but it will fail non-deterministically once it does. The normalization re-sorts both sides with the same comparator so we're asserting the data is identical, not just that both sides happen to agree on tie-breaking order.

File: src/server/logs/victorialogs/provider.ts (hunk 2 above)
Comment (bellorr) [reply src/server/logs/victorialogs/provider.ts]: Both fixed in 74846cd:

- **Pagination**: `getQueryLogs`, `getTopDomains`, and `getTopClients` now push `| offset N | limit N` directly into the VL query, with a separate `| stats count()` query for `totalCount`. No more fetching all rows and slicing in Node.

- **Secondary sort**: VL does support multi-field sort. Added tie-breakers across all sorted queries — `sort by (_time desc, question_name asc)` in `getQueryLogs`, `sort by (count desc, question_name asc)` / `sort by (total desc, client_names asc)` in the top lists, and similarly for `getQueryTypesBreakdown`, `searchDomains`, and `searchClients`.

File: src/server/logs/victorialogs/provider.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bellorr) [reply src/server/logs/victorialogs/provider.ts]: Confirmed valid and fixed in 74846cd. Added `fillVlTimeBuckets` which pre-generates the full bucket sequence for the requested range (aligned to VL's UTC clock boundaries using `Math.floor(startTime / interval) * interval`), then merges VL rows in by key lookup. Quiet periods now produce zero-value buckets instead of gaps, matching the behaviour of `fillTimeBuckets` in the SQL provider. Also added `toVlKey` to normalize timestamps for the lookup — VL returns `YYYY-MM-DDTHH:MM:SSZ` (no milliseconds) while `Date.toISOString()` produces `YYYY-MM-DDTHH:MM:SS.mmmZ`, which would have caused all keys to miss without stripping the fractional part.

File: src/server/logs/victorialogs/provider.ts [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    return lines.map((line) => {
+      try {
+        return JSON.parse(line) as Record<string, string>;
+      } catch {
+        throw new Error(
+          `VictoriaLogs: failed to parse response line (${line.length} chars)`,
+        );
+      }
+    });
+  }
```
[the remark this answers is not in the document]
Comment (bellorr) [reply src/server/logs/victorialogs/provider.ts:128]: Fixed in 526385c. Wrapped the entire `queryRaw` body in an outer `try/catch` that rethrows with `new Error("VictoriaLogs: queryRaw failed", { cause: error })`, preserving the original error via the `cause` chain. The inner per-line throw (parse errors) is unchanged.

**Comment by bellorr** [comment]
@GabeDuarteM Request for you to review this PR again whenever you have the time. I rebased it to the latest release you have.

File: src/server/logs/__tests__/configure-container-runtime.ts [hunk 7]
```diff
@@ -42,17 +42,30 @@ if (!process.env["DOCKER_HOST"]) {
...
 
-    if (uid !== undefined) {
-      const podmanSocket = `/run/user/${uid}/podman/podman.sock`;
+    if (isSocketAccessible(colimaSocket)) {
+      process.env["DOCKER_HOST"] = `unix://${colimaSocket}`;
+      // Ryuk bind-mounts the Docker socket into its own container using the path
+      // from DOCKER_HOST. The Colima socket lives at a non-standard host path that
+      // cannot be resolved from inside a container, so Ryuk fails to start.
+      // Disable it; containers are cleaned up by the afterAll hooks instead.
+      process.env["TESTCONTAINERS_RYUK_DISABLED"] = "true";
```
Comment (GabeDuarteM) [inline src/server/logs/__tests__/configure-container-runtime.ts]: That avoids the socket bind-mount problem, but it also removes Testcontainers resource reaper. If `beforeAll` fails partway through setupProviders() or the process is killed, containers can leak. `afterAll` does cleans up, but that only helps if setup reaches the point where cleanup is assigned and Vitest exits normally.

For colima specifically, testcontainers seem to [recommend this](https://node.testcontainers.org/supported-container-runtimes/#colima):

```env
export DOCKER_HOST=unix://${HOME}/.colima/default/docker.sock
export TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/var/run/docker.sock
```

we could do the same with this:

```ts
      process.env["DOCKER_HOST"] = `unix://${colimaSocket}`;
      process.env["TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE"] ??=
        "/var/run/docker.sock";
```

can you check if this works in your setup?

File: src/server/logs/__tests__/provider-conformance.test.ts (hunk 1 above)
Comment (GabeDuarteM) [reply src/server/logs/__tests__/provider-conformance.test.ts]: I checked the VictoriaLogs docs and seems like the format pipe supports lowercase field formatting with lc:, and sort supports ordering by fields. Can we use that instead? 

```ts
const QUESTION_NAME_SORT_KEY = "__question_name_sort";
const CLIENT_NAMES_SORT_KEY = "__client_names_sort";
...
      this.queryRaw(
        `${base} | stats by (question_name) count() as count | format "<lc:question_name>" as ${QUESTION_NAME_SORT_KEY} | sort by (count desc, ${QUESTION_NAME_SORT_KEY} asc) | offset ${offset} | limit ${limit}`,
        { start },
      ),
```

https://docs.victoriametrics.com/victorialogs/logsql/#format-pipe
https://docs.victoriametrics.com/victorialogs/logsql/#sort-pipe

I'd really like to avoid changing the test to fit the provider and risk having a legit regression undetected, instead of actually having the provider performing the same way as all others...

**Review by GabeDuarteM — COMMENTED** [review]
Hey, sorry for the delay, I just really wanted to take a deeper look myself into this due to the change in provider conformance test. 

Even though I would really much like adding new providers like VictoriaLogs, I'm honestly not comfortable in merging it if it requires changing the provider conformance test to fit the new provider. It would defeat the purpose of the test in the first place, since its idea is to make sure they all behave more or less the same way.

I've added a couple comments with maybe another solution for this, can you please check if they also work for you?

**Comment by bellorr** [comment]
@GabeDuarteM 

Thanks for the detailed feedback and for taking the time to dig into the VL docs yourself.

You were right about the conformance test. Modifying it to accommodate a provider rather than fixing the provider to match defeats the whole purpose. The fix was to use VL's `format "<lc:field>"` pipe to produce a lowercase sort key, which makes the secondary sort case-insensitive and consistent with MySQL's default collation and CSV's `localeCompare`.

Also addressed the Colima/Ryuk issue using `TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE` as the testcontainers docs recommend.

All 327 tests passing.
